File: Data Structures & Algorithms/rotting-fruit/submission-2.js
class Solution {
    /**
     * @param {number[][]} grid
     * @return {number}
     */
    orangesRotting(grid) {
        let queue = []
        let ROWS = grid.length, COLS = grid[0].length
        let fresh = 0
        let time = 0
        let directions = [[0, 1], [0, -1], [1, 0], [-1, 0]]

        for(let i = 0; i < ROWS; i++) {
            for(let j = 0; j < COLS; j++) {
                if(grid[i][j] === 1) {
                    fresh++
                } else if(grid[i][j] === 2) {
                    queue.push([i,j])
                }
            }
        }

        while(queue.length > 0 && fresh > 0) {
            let size = queue.length
            for(let i = 0; i < size; i++) {
                let [r, c] = queue.shift();

                for(let dir of directions) {
                    let [dr, dc] = dir

                    if(Math.min(r + dr, c + dc) < 0 ||
                        r + dr === ROWS ||
                        c + dc === COLS ||
                        grid[r + dr][c + dc] === 2 ||
                        grid[r + dr][c + dc] === 0
                    ) {
                        continue;
                    }

                    grid[r + dr][c + dc] = 2
                    fresh--;
                    queue.push([r + dr, c + dc])
                }
            }
            time++;
        }

        return fresh === 0 ? time : -1 
    }
}

File: Data Structures & Algorithms/islands-and-treasure/submission-2.py
class Solution:
    def islandsAndTreasure(self, grid: List[List[int]]) -> None:
        ROWS, COLS = len(grid), len(grid[0])
        queue = deque()
        visited = set()

        def addCell(r, c):
            if (min(r, c) < 0 or r == ROWS or c == COLS or
                (r, c) in visited or grid[r][c] == -1):
                return
            visited.add((r, c))
            queue.append([r, c])
        
        for row in range(ROWS):
            for col in range(COLS):
                if grid[row][col] == 0:
                    queue.append([row, col])
                    visited.add((row, col))
    
        neighbors = [[1,0],[-1,0],[0,1],[0,-1]]
        level = 0
        while queue:
            for i in range(len(queue)):
                r, c = queue.popleft()
                grid[r][c] = level
                for dr, dc in neighbors:
                    if (min(r + dr, c + dc) < 0 or
                        r + dr == ROWS or c + dc == COLS or
                        (r + dr, c + dc) in visited or grid[r + dr][c + dc] == -1):
                            continue
                    visited.add((r + dr, c + dc))
                    queue.append([r + dr, c + dc])
            level += 1